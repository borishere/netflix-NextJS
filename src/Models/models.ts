import * as yup from 'yup';

export interface Igenre {
  name: string;
  active: boolean;
}

export interface ImovieBase {
  title: string;
  overview: string;
  release_date?: string;
  poster_path: string;
  vote_average?: number;
  genres: string[];
  runtime: number;
  tagline?: string;
  vote_count?: number;
  budget?: number;
  revenue?: number;
}

export interface Imovie extends ImovieBase {
  id: number;
}

export type TNullableMovie = Imovie | null;

export interface IMoviesResponse {
  data: Imovie[];
  total: number;
  offset: number;
  limit: number;
}

export interface IMoviesState {
  sortBy: string;
  filter?: string[];
}

export interface IGetMoviesArgs {
  sortBy?: string;
  sortOrder?: 'asc' | 'desc'
  filter?: string | string[]
  search?: string,
  searchBy?: string;
  movie?: string;
}

export const validationSchema = yup.object().shape({
  title: yup.string().required(),
  release_date: yup.date(),
  vote_average: yup.number(),
  genres: yup.array().of(yup.string()).required(),
  poster_path: yup.string().url().required(),
  overview: yup.string().required(),
  runtime: yup.number().min(0).integer().required()
});