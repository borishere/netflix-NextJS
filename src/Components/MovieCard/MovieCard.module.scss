.movie-item {
  position: relative;
  margin: 30px 55px 0 0;
  cursor: pointer;

  .context-btn {
    position: absolute;
    top: 20px;
    right: 20px;
    cursor: pointer;
  }

  .movie-content {
    display: flex;
    flex-direction: column;

    .movie-content-base {
      display: flex;
      flex-direction: row;
      justify-content: space-between;
      align-items: center;
    }
  }
}