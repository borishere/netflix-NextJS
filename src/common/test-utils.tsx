import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ModalContext } from '../Context/ModalContext';
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import { createMemoryHistory } from 'history';
import { moviesApi } from '../services/movies';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import moviesReducer from '../app/moviesSlice';
import { configureStore } from '@reduxjs/toolkit';

const history = createMemoryHistory();


const AllTheProviders: FC = ({ children }) => {
  return (
    <Router location='/' navigator={history}>
      <Provider store={store}>
        <ModalContext.Provider value={{}}>
          {children}
        </ModalContext.Provider>
      </Provider>
    </Router>
  );
};

function renderWithStore(
  ui: ReactElement,
  {
    store = configureStore({
      reducer: {
        [moviesApi.reducerPath]: moviesApi.reducer,
        movies: moviesReducer

      },
      middleware: (getDefaultMiddleware: any) =>
        getDefaultMiddleware().concat(moviesApi.middleware)
    }),
    ...renderOptions
  } = {}
) {
  setupListeners(store.dispatch);

  function Wrapper({ children }: any) {
    const history = createMemoryHistory();

    return (
      <Provider store={store}>
        <Router location='/' navigator={history}>
          {children}
        </Router>
      </Provider>
    );
  }

  store.dispatch(moviesApi.util.resetApiState());

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

const root = document.createElement('div');
root.id = 'root';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, container: document.body.appendChild(root), ...options });

export * from '@testing-library/react';
export { customRender, renderWithStore };