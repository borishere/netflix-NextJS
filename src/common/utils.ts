export function addParamToExistsSearchParams(key: string, value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.set(key, value);

  return params;
}

export function deleteParamFromExistsSearchParams(value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.delete(value);

  return params;
}

export const BASE_API_URL: string = 'http://localhost:4000/';