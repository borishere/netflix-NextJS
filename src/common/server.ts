import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { BASE_API_URL } from './utils';
import mockData from './__mocks__/movies.json';

let movies: any[] = Array.from(mockData);

export const server = setupServer(
  rest.get(`${BASE_API_URL}movies`, (req, res, ctx) => {
    return res(ctx.json(movies.slice(0, 2)));
  }),

  rest.get(`${BASE_API_URL}movies:id`, (req, res, ctx) => {
    const { id } = req.params;
    const movie = movies.find((movie) => movie.id === id);

    return res(ctx.json(movie));
  }),

  rest.post(`${BASE_API_URL}movies`, (req, res, ctx) => {
    movies.push(req.body);

    return res(ctx.status(200));
  }),

  rest.put(`${BASE_API_URL}movies`, (req, res, ctx) => {
    const requestMovie: any = req.body;

    const movieIndex = movies.findIndex((m) => m.id === requestMovie.id);

    if (movieIndex < 0) {
      return res(ctx.status(404));
    }

    movies[movieIndex] = requestMovie;

    return res(ctx.status(200));
  }),

  rest.delete(`${BASE_API_URL}movies/:id`, (req, res, ctx) => {
    const { id } = req.params;

    movies = movies.filter((m) => m.id !== id);

    return res(ctx.status(200));
  })
);
