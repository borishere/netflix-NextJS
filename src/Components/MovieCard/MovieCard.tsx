import React, { FC, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { ContextMenu } from '../ContextMenu/ContextMenu';
// @ts-ignore
import contextIcon from '../../Images/item-context.svg';
import { Imovie } from '../../Models/models';
import { addParamToExistsSearchParams } from '../../common/utils';
import './style.scss';

interface Props {
  movie: Imovie;
}

export const MovieCard: FC<Props> = ({ movie }) => {
  const [showContextMenu, setShowContextMenu] = useState<boolean>(false);
  const navigate = useNavigate();

  const onMovieSelect = () => {
    const params = addParamToExistsSearchParams('movie', movie.id.toString());

    navigate({ search: params.toString() });
  };

  return (
    <li className='movie-item' onClick={onMovieSelect}>
      <img
        className='context-btn'
        src={contextIcon}
        onClick={() => setShowContextMenu(!showContextMenu)}
      />
      <img src={movie?.poster_path} />
      {showContextMenu && (
        <ContextMenu
          setShow={setShowContextMenu}
          movie={movie}
        />
      )}
      <div className='movie-content'>
        <div className='movie-content-base'>
          <h2 className='movie-title'>{movie?.title}</h2>
          <span className='movie-release-date'>{movie?.release_date}</span>
        </div>
        <span className='movie-genres'>{movie?.genres?.map((genre, i) => `${genre}${i < movie.genres.length - 1 ? ', ' : ''}`)}</span>
      </div>
    </li>
  );
};