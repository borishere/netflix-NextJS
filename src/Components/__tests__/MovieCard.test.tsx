import React from 'react';
import { screen, customRender } from '../../common/test-utils';
import '@testing-library/jest-dom';
import { Imovie } from '../../Models/models';
import { MovieCard } from '../MovieCard/MovieCard';

describe('MovieCard', () => {
  const movie: Imovie = {
    id: 76338,
    title: 'Thor: The Dark World',
    tagline: 'Delve into the darkness',
    vote_average: 6.7,
    vote_count: 6522,
    release_date: '2013-10-29',
    poster_path: 'https://image.tmdb.org/t/p/w500/bnX5PqAdQZRXSw3aX3DutDcdso5.jpg',
    overview: 'Thor fights to restore order across the cosmosâ€¦ but an ancient race led by the vengeful Malekith returns to plunge the universe back into darkness. Faced with an enemy that even Odin and Asgard cannot withstand, Thor must embark on his most perilous and personal journey yet, one that will reunite him with Jane Foster and force him to sacrifice everything to save us all.',
    budget: 170000000,
    revenue: 644571402,
    genres: [
      'documentary',
      'comedy'
    ],
    runtime: 112
  };

  test('Renders', () => {
    const { asFragment } = customRender(
      <MovieCard movie={movie} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('Renders movie values', () => {
    customRender(
      <MovieCard movie={movie} />
    );

    expect(screen.getByText(movie.title)).toBeInTheDocument();
    expect(screen.getByText(movie.release_date as string)).toBeInTheDocument();
  });
});