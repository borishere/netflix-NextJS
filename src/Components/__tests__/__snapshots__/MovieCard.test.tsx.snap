// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard Renders 1`] = `
<DocumentFragment>
  <li
    class="movie-item"
  >
    <img
      class="context-btn"
      src="test-file-stub"
    />
    <img
      src="https://image.tmdb.org/t/p/w500/bnX5PqAdQZRXSw3aX3DutDcdso5.jpg"
    />
    <div
      class="movie-content"
    >
      <div
        class="movie-content-base"
      >
        <h2
          class="movie-title"
        >
          Thor: The Dark World
        </h2>
        <span
          class="movie-release-date"
        >
          2013-10-29
        </span>
      </div>
      <span
        class="movie-genres"
      >
        documentary, comedy
      </span>
    </div>
  </li>
</DocumentFragment>
`;
